v-next/hardhat-verify/src/internal/blockscout.ts

```diff
@@ -0,0 +1,359 @@
+        {
+          name: this.name,
+          url: this.apiUrl,
+          statusCode: response.statusCode,
+          errorMessage: responseBody.result,
+        },
+      );
+    }
+
+    const blockscoutResponse = new BlockscoutVerificationResponse(responseBody);
+
+    if (blockscoutResponse.isBytecodeMissingInNetworkError()) {
+      throw new HardhatError(
+        HardhatError.ERRORS.HARDHAT_VERIFY.GENERAL.CONTRACT_VERIFICATION_MISSING_BYTECODE,
+        {
+          url: this.apiUrl,
+          address: contractAddress,
+        },
+      );
+    }
+
+    if (blockscoutResponse.isAlreadyVerified()) {
+      throw new HardhatError(
+        HardhatError.ERRORS.HARDHAT_VERIFY.GENERAL.CONTRACT_ALREADY_VERIFIED,
+        {
+          contract: contractName,
+          address: contractAddress,
+        },
+      );
+    }
+
+    if (!blockscoutResponse.isOk()) {
+      throw new HardhatError(
+        HardhatError.ERRORS.HARDHAT_VERIFY.GENERAL.CONTRACT_VERIFICATION_REQUEST_FAILED,
+        { message: blockscoutResponse.message },
+      );
+    }
+
+    return blockscoutResponse.message;
+  }
+
+  public async pollVerificationStatus(
+    guid: string,
+    contractAddress: string,
+    contractName: string,
+  ): Promise<{
+    success: boolean;
+    message: string;
+  }> {
+    let response: HttpResponse;
+    let responseBody: BlockscoutResponse | undefined;
+    try {
+      response = await getRequest(
+        this.apiUrl,
+        {
+          queryParams: {
+            module: "contract",
+            action: "checkverifystatus",
+            guid,
+          },
+        },
+        this.#dispatcher,
+      );
+      responseBody =
+        /* eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+        -- Cast to BlockscoutResponse because that's what we expect from the API
+        TODO: check if the API returns a different type and throw an error if it does */
+        (await response.body.json()) as BlockscoutResponse;
+    } catch (error) {
+      ensureError(error);
+
+      throw new HardhatError(
+        HardhatError.ERRORS.HARDHAT_VERIFY.GENERAL.EXPLORER_REQUEST_FAILED,
+        {
+          name: this.name,
+          url: this.apiUrl,
+          errorMessage:
+            error.cause instanceof Error ? error.cause.message : error.message,
+        },
+      );
+    }
+
+    const isSuccessStatusCode =
+      response.statusCode >= 200 && response.statusCode <= 299;
+    if (!isSuccessStatusCode) {
+      // TODO: we should consider throwing EXPLORER_REQUEST_FAILED here too
+      throw new HardhatError(
+        HardhatError.ERRORS.HARDHAT_VERIFY.GENERAL.EXPLORER_REQUEST_STATUS_CODE_ERROR,
+        {
+          name: this.name,
+          url: this.apiUrl,
+          statusCode: response.statusCode,
+          errorMessage: responseBody.result,
+        },
+      );
+    }
+
+    const blockscoutResponse = new BlockscoutVerificationStatusResponse(
```

Comment:
Nit: no need to change - or even agree.
I suspect I would have dodged the class here and done a straight transform to pure data, rather than wrapping and deferring the calculation of the results.
This status response class feels like a pure data carrier.

Reply:
I agree. The same applies to the `EtherscanVerificationStatusResponse` class. I intend to clean this up in the future.